apps/ejabberd/src/ejabberd_router.erl

```diff
@@ -295,50 +311,65 @@ code_change(_OldVsn, State, _Extra) ->
 %%% Internal functions
 %%--------------------------------------------------------------------
 
-do_route(OrigFrom, OrigTo, OrigPacket) ->
-    ?DEBUG("route~n\tfrom ~p~n\tto ~p~n\tpacket ~p~n",
-           [OrigFrom, OrigTo, OrigPacket]),
-    %% Filter globally
-    case ejabberd_hooks:run_fold(filter_packet,
-                                 {OrigFrom, OrigTo, OrigPacket}, []) of
-        {From, To, Packet} ->
-            LDstDomain = To#jid.lserver,
-            case mnesia:dirty_read(external_component, LDstDomain) of
-                [] ->
-                    case mnesia:dirty_read(route, LDstDomain) of
-                        [] ->
-                            ejabberd_s2s:route(From, To, Packet);
-                        [#route{handler=Handler}] ->
-                            do_local_route(OrigFrom, OrigTo, OrigPacket,
-                                           LDstDomain, Handler)
-                    end;
-                [#external_component{handler = Handler}] ->
-                    do_local_route(OrigFrom, OrigTo, OrigPacket,
-                                   LDstDomain, Handler)
-            end;
-        drop ->
-            ejabberd_hooks:run(xmpp_stanza_dropped,
-                               OrigFrom#jid.lserver,
-                               [OrigFrom, OrigTo, OrigPacket]),
-            ok
-    end.
-
-do_local_route(OrigFrom, OrigTo, OrigPacket, LDstDomain, Handler) ->
-    %% Filter locally
-    case ejabberd_hooks:run_fold(filter_local_packet, LDstDomain,
-                                 {OrigFrom, OrigTo, OrigPacket}, []) of
-        {From, To, Packet} ->
-            case Handler of
-                {apply_fun, Fun} ->
-                    Fun(From, To, Packet);
-                {apply, Module, Function} ->
-                    Module:Function(From, To, Packet)
-            end;
+routing_modules_list() ->
+    %% this is going to be compiled on startup from settings
+    mod_routing_machine:get_routing_module_list().
+
+compile_routing_module() ->
+    Mods = ejabberd_config:get_local_option(routing_modules),
```

Comment:
I think it's a good practice to handle `undefined` values on the system border as early as we hit them instead of allowing them to be passed around (they're the Erlang equivalent of nullable types eschewed in general). Ideally, `get_local_option` would take the default as another parameter to return in case it doesn't find the option ([it does in ejabberd](https://github.com/processone/ejabberd/blob/a9e50468b613f8b988e91e65b79ad155b65c622f/src/ejabberd_config.erl#L759), though `Default` was introduced after we forked). In this case, I'd do:

```erlang
case ejabberd_config:get_local_option(routing_modules) of
    undefined -> default_routing_modules();
    Defined -> Defined
end
```

Later on, we can make `make_routing_module_source/1` take a list and actually verify that it's a list with `when is_list/1`.